java/java.hints/src/org/netbeans/modules/java/hints/jdk/ConvertToNestedRecordPattern.java

```diff
@@ -0,0 +1,261 @@
+            return null;
+        }
+        TreePath t = ctx.getPath();
+        if (!t.getParentPath().getLeaf().getKind().equals(Tree.Kind.INSTANCE_OF)) {
+            return null;
+        }
+        Set<String> recordPatternVarSet = new HashSet<>();
+        Map<PatternTree, List<PatternTree>> recordComponentMap = new LinkedHashMap<>();
+        DeconstructionPatternTree recordPattern = (DeconstructionPatternTree) t.getLeaf();
+        recordComponentMap = findNested(recordPattern, recordComponentMap);
+
+        for (PatternTree p : recordComponentMap.keySet()) {
+            BindingPatternTree bTree = (BindingPatternTree) p;
+            recordPatternVarSet.add(bTree.getVariable().getName().toString());
+        }
+        while (t != null && t.getLeaf().getKind() != Tree.Kind.IF) {
+            t = t.getParentPath();
+        }
+        Set<TreePath> convertPath = new HashSet<>();
+        List<String> localVarList = new ArrayList<>();
+        Map<String, List<UserVariables>> userVars = new HashMap<>();
+        new TreePathScanner<Void, Void>() {
+
+            @Override
+            public Void visitVariable(VariableTree node, Void p) {
+                localVarList.add(node.getName().toString());
+                Map<String, TreePath> outerVariables = new HashMap<>();
+                Map<String, String> innerVariables = new HashMap<>();
+                List<UserVariables> nList = new ArrayList<>();
+                boolean match = MatcherUtilities.matches(ctx, getCurrentPath(), "$type $var1 = $expr3.$meth1()", outerVariables, new HashMap<>(), innerVariables);
+
+                if (match && recordPatternVarSet.contains(outerVariables.get("$expr3").getLeaf().toString())) {
+                    String expr3 = outerVariables.get("$expr3").getLeaf().toString();
+                    nList.clear();
+                    if (userVars.get(expr3) != null) {
+                        nList = userVars.get(expr3);
+                    }
+                    nList.add(new UserVariables(innerVariables.get("$var1"), innerVariables.get("$meth1")));
+                    userVars.put(expr3, nList);
+                    convertPath.add(getCurrentPath());
+                }
+                return super.visitVariable(node, p);
+            }
+        }.scan(t, null);
+        if (!convertPath.isEmpty()) {
+            Fix fix = new FixImpl(ctx.getInfo(), ctx.getPath(), convertPath, localVarList, userVars).toEditorFix();
+            return ErrorDescriptionFactory.forName(ctx, ctx.getPath(), Bundle.ERR_ConvertToNestedRecordPattern(), fix);
+        }
+        return null;
+    }
+
+    private static Map<PatternTree, List<PatternTree>> findNested(PatternTree pTree, Map<PatternTree, List<PatternTree>> recordComponentMap) {
+        if (pTree instanceof BindingPatternTree) {
```

Comment:
Better use `pTree.getKind()`, and switch over that kind, having a `default` with an exception. `instanceof` is not necessarily a safe idiom when working with `com.sun.source.tree`.

Reply:
handled